rules/manager.go

```diff
@@ -407,12 +422,133 @@ func (g *Group) Eval(ctx context.Context, ts time.Time) {
 	}
 }
 
+// RestoreForState restores the 'for' state of the alerts
+// by looking up last ActiveAt from storage.
+func (g *Group) RestoreForState(ts time.Time) {
+	maxtMS := int64(model.TimeFromUnixNano(ts.UnixNano()))
+	// We allow restoration only if alerts were active before some certain time.
+	mint := ts.Add(-g.opts.OutageTolerance)
+	mintMS := int64(model.TimeFromUnixNano(mint.UnixNano()))
+	q, err := g.opts.TSDB.Querier(g.opts.Context, mintMS, maxtMS)
+	if err != nil {
+		level.Error(g.logger).Log("msg", "Failed to get Querier", "err", err)
+	}
+
+	for _, rule := range g.Rules() {
+		alertRule, ok := rule.(*AlertingRule)
+		if !ok {
+			continue
+		}
+
+		alertHoldDuration := alertRule.HoldDuration()
+
+		alertRule.ForEachActiveAlert(func(a *Alert) {
+
+			smpl := alertRule.forStateSample(a, time.Now(), 0)
+			var matchers []*labels.Matcher
+			for _, l := range smpl.Metric {
+				mt, _ := labels.NewMatcher(labels.MatchEqual, l.Name, l.Value)
+				matchers = append(matchers, mt)
+			}
+
+			sset, err := q.Select(nil, matchers...)
+			if err != nil {
+				level.Error(g.logger).Log("msg", "Failed to restore 'for' state",
+					labels.AlertName, alertRule.Name(), "stage", "Select", "err", err)
+				return
+			}
+
+			seriesFound := false
+			var s storage.Series
+			for sset.Next() {
+				if len(sset.At().Labels()) == len(smpl.Metric) {
+					s = sset.At()
+					seriesFound = true
+					break
+				}
+			}
+
+			if !seriesFound {
+				return
+			}
+
+			// Series found for the 'for' state.
+			var v float64
+			var t int64
+			it := s.Iterator()
+			for it.Next() {
+				t, v = it.At()
+			}
+			if value.IsStaleNaN(v) { // Alert was not active.
+				return
+			}
+
+			// v = the latest ActiveAt of the alert.
+			restoredActiveAt := time.Unix(int64(v), 0)
+			crashTime := time.Unix(t/1000, 0)
+			if crashTime.Sub(restoredActiveAt) >= alertHoldDuration {
+				// The alert was firing when it crashed. Hence we need not wait for grace
+				// period, we can immediatly make it fire.
+				a.ActiveAt = restoredActiveAt
+				a.State = StateFiring
+				return
+			}
+
+			// This makes sure that the `for` period of the alerting rule is
+			// greater then the grace period. We take application of grace period
+			// into consideration only if this condition is satisfied.
+			shouldWaitForGracePeriod := alertHoldDuration > g.opts.ForGracePeriod
+
+			// restoredActiveAt               ts (current)             firing time = restoredActiveAt.Add(alertHoldDuration)
+			//     |-------------------------------|-------------------------------|
+			//
+			//     ^-----------------------alertHoldDuration-----------------------^
+			//
+			//                                     ^-------------------------------^
+			// 			    					restoredActiveAt.Add(alertHoldDuration).Sub(ts)
+			//
+			// Hence `restoredActiveAt.Add(alertHoldDuration).Sub(ts)` is the time left
```

Comment:
I think I see where the disconnect is. If an alert has a for 6h, I think it must spend at least 6h firing. If the Prometheus is down for half an hour, it now needs to spend at least 6h+30m+10m firing. This will reduce false positives due to alerts resolving while Prometheus is down.

Reply:
With firing do you mean pending time? I did not quite follow what you meant.

Reply:
Yes, I meant pending.

Reply:
So 6h+30m+10m is a problem?

Reply:
According to the implementation, it will be pending for another 3h30m before firing, as the total time after it got active would be as `2h + 30m (down) + 3h30m = 6h`.  Or do we want it to be pending for more 4h making total time from when it was active to be `6h30m`?

Reply:
The total time there should be 6h30m or 6h40m depending on how we deal with the 10m in that case.

Reply:
Okay, now I get why you were behind the timestamp. I will make that change.

Reply:
I have make that change. Now in the above case the total time will be 6h30m or 6h40m depending on grace period. The down time wont be considered in pending time.